### DISCUSSION TITLE
Is there a MSTest TestContext.AddresultFile() alternative?

### DISCUSSION THREAD

**Discussion by baumheld** [thomhurst/TUnit#1276] [Q&A] [answered]
In MSTest, I use [TestContext.AddResultFile(String)](https://learn.microsoft.com/dotnet/api/microsoft.visualstudio.testtools.unittesting.testcontext.addresultfile) to attach custom logiles, zip-files, screenshots from UI-Tests to testcontexts of failed tests.

I'd really like to switch over to TUnit after reading your documentation. Since you offer a static TestContext singleton, I was wondering if one could somehow attach custom files when also using the Microsoft TRX nuget. These custom files then get uploaded to Azure DevOps automatically when publishing test results and can be viewed in their Azure Devops unit test explorer.

My question is: Can I mimic this feature in TUnit?

**Comment by thomhurst** [comment, marked as answer]
The test context exposes an add artifact method. Would that suffice? If not let me know what you'd need.

**Reply by baumheld** [reply]
That worked, thx :)

For future readers:
When using the TRX nuget the `AddArtifact()` method added a `ResultFiles` entry which then can be uploaded to Azure DevOps and their test results view via standard `PublishTestResults@2` task

### .trx file snippet
```xml
[...]
  <Results>
    <UnitTestResult [...]>
      <ResultFiles>
        <ResultFile path="C:\foo\bar.log" />
      </ResultFiles>
    </UnitTestResult>
  </Results>
[...]
```


### .cs file
```csharp

public class TestClass
{
    [Test]
    public void AddArtifact_Test()
    {
        var logfile = new FileInfo(@"C:\foo\bar.log");
        var artifact = new Artifact() { File = logfile, DisplayName = logfile.Name };
        TestContext.Current?.AddArtifact(artifact);
    }
}
```

### .csproj file

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net9.0</TargetFramework>
    <ImplicitUsings>disable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Testing.Extensions.TrxReport" Version="1.4.3" />
    <PackageReference Include="TUnit" Version="0.3.43" />
  </ItemGroup>

</Project>
```
